lib/integrations/awsoidc/eks_list_clusters.go

```diff
@@ -0,0 +1,186 @@
+type ListEKSClustersResponse struct {
+	// Servers contains the page of Servers.
+	Clusters []EKSCluster
+
+	// NextToken is used for pagination.
+	// If non-empty, it can be used to request the next page.
+	NextToken string
+
+	// ClusterFetchingErrors contains errors for fetching detailed information about specific cluster, if any happened.
+	ClusterFetchingErrors map[string]error
+}
+
+// NewListEKSClustersClient creates a new ListEKSClusters client using AWSClientRequest.
+func NewListEKSClustersClient(ctx context.Context, req *AWSClientRequest) (ListEKSClustersClient, error) {
+	clt, err := newEKSClient(ctx, req)
+	return clt, trace.Wrap(err)
+}
+
+// ListEKSClustersClient describes the required methods to List EKS clusters using a 3rd Party API.
+type ListEKSClustersClient interface {
+	// ListClusters lists the EKS clusters.
+	ListClusters(ctx context.Context, params *eks.ListClustersInput, optFns ...func(*eks.Options)) (*eks.ListClustersOutput, error)
+
+	// DescribeCluster returns detailed information about an EKS cluster.
+	DescribeCluster(ctx context.Context, params *eks.DescribeClusterInput, optFns ...func(*eks.Options)) (*eks.DescribeClusterOutput, error)
+}
+
+// concurrentEKSFetchingLimit is a limit of how many clusters we are trying to describe concurrently, after receiving a list of clusters.
+const concurrentEKSFetchingLimit = 5
+
+// ListEKSClusters calls the following AWS API:
+// https://docs.aws.amazon.com/eks/latest/APIReference/API_ListClusters.html - to list available EKS clusters
+// https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeCluster.html - to get more detailed information about
+// the each cluster in the list we received.
+// It returns a list of EKS clusters with detailed information about them.
+func ListEKSClusters(ctx context.Context, clt ListEKSClustersClient, req ListEKSClustersRequest) (*ListEKSClustersResponse, error) {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	listEKSClusters := &eks.ListClustersInput{
+		Include: nil,
+	}
+	if req.NextToken != "" {
+		listEKSClusters.NextToken = &req.NextToken
+	}
+	eksClusters, err := clt.ListClusters(ctx, listEKSClusters)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var mu sync.Mutex
+	ret := &ListEKSClustersResponse{
+		NextToken:             aws.ToString(eksClusters.NextToken),
+		ClusterFetchingErrors: map[string]error{},
+	}
+
+	group, groupCtx := errgroup.WithContext(ctx)
+	group.SetLimit(concurrentEKSFetchingLimit)
+
+	ret.Clusters = make([]EKSCluster, 0, len(eksClusters.Clusters))
+	for _, clusterName := range eksClusters.Clusters {
+		clusterName := clusterName
+		if clusterName == "" {
+			continue
+		}
+		group.Go(func() error {
+			eksClusterInfo, err := clt.DescribeCluster(groupCtx, &eks.DescribeClusterInput{
+				Name: aws.String(clusterName),
+			})
+
+			mu.Lock()
+			defer mu.Unlock()
+
+			if err != nil {
+				ret.ClusterFetchingErrors[clusterName] = err
+			} else {
```

Comment:
Can we return immediately to reduce the nesting level?
```suggestion
			if err != nil {
				ret.ClusterFetchingErrors[clusterName] = err
				return nil
			}
```